File: worker/worker-boxlite/src/capability/terminal_resource.rs
use std::sync::Arc;
use std::sync::atomic::Ordering;
use std::time::Instant;

use boxlite::BoxCommand;
use futures::StreamExt;
use serde::{Deserialize, Serialize};

use super::CapabilityError;
use super::terminal_exec::SessionManager;
use crate::config::Config;

const PROBE_SCRIPT: &str = r#"
import argparse
import base64
import json
import mimetypes
import os
import sys

parser = argparse.ArgumentParser()
parser.add_argument("--action", choices=["validate", "read"], default="validate")
parser.add_argument("--file-path", required=True)
parser.add_argument("--max-read-bytes", type=int, required=True)
args = parser.parse_args()

target = args.file_path
if not os.path.exists(target):
    print(json.dumps({"error": "file_not_found", "message": "file not found"}))
    sys.exit(10)
if os.path.isdir(target):
    print(json.dumps({"error": "path_is_directory", "message": "path is directory"}))
    sys.exit(11)

size_bytes = os.path.getsize(target)
mime_type, _ = mimetypes.guess_type(target)
if not mime_type:
    mime_type = "application/octet-stream"

if args.action == "validate":
    print(json.dumps({"mime_type": mime_type, "size_bytes": size_bytes}))
    sys.exit(0)

limit = args.max_read_bytes
if size_bytes > limit:
    print(json.dumps({
        "error": "file_too_large",
        "message": "file exceeds read limit",
        "mime_type": mime_type,
        "size_bytes": size_bytes,
    }))
    sys.exit(12)

with open(target, "rb") as fh:
    content = fh.read(limit + 1)
if len(content) > limit:
    print(json.dumps({
        "error": "file_too_large",
        "message": "file exceeds read limit",
        "mime_type": mime_type,
        "size_bytes": len(content),
    }))
    sys.exit(12)

print(json.dumps({
    "mime_type": mime_type,
    "size_bytes": len(content),
    "blob": base64.b64encode(content).decode("ascii"),
}))
"#;

// ---------- Request / Response ----------

#[derive(Deserialize)]
struct TerminalResourceRequest {
    session_id: String,
    file_path: String,
    #[serde(default)]
    action: Option<String>,
}

#[derive(Serialize)]
struct TerminalResourceResponse {
    session_id: String,
    file_path: String,
    mime_type: String,
    size_bytes: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    blob: Option<String>,
}

#[derive(Deserialize)]
struct ProbeResult {
    #[serde(default)]
    error: Option<String>,
    #[serde(default)]
    message: Option<String>,
    #[serde(default)]
    mime_type: Option<String>,
    #[serde(default)]
    size_bytes: i64,
    #[serde(default)]
    blob: Option<String>,
}

// ---------- Handler ----------

pub async fn handle_terminal_resource(
    session_mgr: &Arc<SessionManager>,
    config: &Config,
    payload: &[u8],
    deadline: Option<Instant>,
) -> Result<Vec<u8>, CapabilityError> {
    let req: TerminalResourceRequest = serde_json::from_slice(payload)
        .map_err(|e| CapabilityError::new("invalid_payload", format!("bad JSON: {e}")))?;

    let session_id = req.session_id.trim().to_string();
    let file_path = req.file_path.trim().to_string();

    if session_id.is_empty() || file_path.is_empty() {
        return Err(CapabilityError::new(
            "invalid_payload",
            "session_id and file_path are required",
        ));
    }

    let action = normalize_action(req.action.as_deref());
    if action.is_empty() {
        return Err(CapabilityError::new(
            "invalid_payload",
            "action must be validate or read",
        ));
    }

    tracing::debug!(
        action = %action,
        session_id_present = true,
        file_path_len = file_path.len(),
        "terminalResource"
    );

    // Lookup session
    let session = session_mgr
        .get_session(&session_id)
        .await
        .ok_or_else(|| CapabilityError::new("session_not_found", "session not found"))?;

    // Acquire busy lock
    if session
        .busy
        .compare_exchange(false, true, Ordering::AcqRel, Ordering::Acquire)
        .is_err()
    {
        return Err(CapabilityError::new(
            "session_busy",
            "session is executing another command",
        ));
    }

    let session_for_guard = session.clone();
    let _busy_guard = scopeguard::guard((), move |_| {
        session_for_guard.busy.store(false, Ordering::Release);
    });

    // Execute probe script
    let limit = config.terminal_output_limit_bytes;
    let exec_fut = async {
        let cmd = BoxCommand::new("python3").args([
            "-c",
            PROBE_SCRIPT,
            "--action",
            &action,
            "--file-path",
            &file_path,
            "--max-read-bytes",
            &limit.to_string(),
        ]);

        let mut execution = session
            .litebox
            .exec(cmd)
            .await
            .map_err(|e| CapabilityError::new("execution_failed", format!("exec: {e}")))?;

        let stdout = collect_stdout(execution.stdout()).await;
        let result = execution
            .wait()
            .await
            .map_err(|e| CapabilityError::new("execution_failed", format!("wait: {e}")))?;

        Ok::<_, CapabilityError>((stdout, result.exit_code))
    };

    let (stdout, exit_code) = match deadline {
        Some(dl) => {
            let timeout = dl.saturating_duration_since(Instant::now());
            tokio::time::timeout(timeout, exec_fut)
                .await
                .map_err(|_| CapabilityError::new("deadline_exceeded", "execution timed out"))?
        }
        None => exec_fut.await,
    }?;

    // Parse probe output
    let trimmed = stdout.trim();
    if trimmed.is_empty() {
        return Err(CapabilityError::new(
            "execution_failed",
            "empty probe output",
        ));
    }

    let probe: ProbeResult = serde_json::from_str(trimmed)
        .map_err(|e| CapabilityError::new("execution_failed", format!("parse probe: {e}")))?;

    // Check for domain errors from probe script
    if let Some(error_code) = &probe.error {
        let code = error_code.trim();
        if !code.is_empty() {
            let message = probe
                .message
                .as_deref()
                .map(|m| m.trim())
                .filter(|m| !m.is_empty())
                .unwrap_or_else(|| default_error_message(code));
            return Err(CapabilityError::new(code, message));
        }
    }

    if exit_code != 0 {
        return Err(CapabilityError::new(
            "execution_failed",
            format!("probe exited with code {exit_code}"),
        ));
    }

    let mime_type = probe
        .mime_type
        .as_deref()
        .map(|m| m.trim())
        .filter(|m| !m.is_empty())
        .unwrap_or("application/octet-stream")
        .to_string();

    let resp = TerminalResourceResponse {
        session_id: session_id.clone(),
        file_path: file_path.clone(),
        mime_type,
        size_bytes: probe.size_bytes,
        blob: if action == "read" {
            Some(probe.blob.unwrap_or_default())
        } else {
            None
        },
    };

    serde_json::to_vec(&resp)
        .map_err(|e| CapabilityError::new("encode_failed", format!("JSON encode: {e}")))
}

fn normalize_action(action: Option<&str>) -> String {
    match action.map(|a| a.trim().to_lowercase()).as_deref() {
        None | Some("") | Some("validate") => "validate".to_string(),
        Some("read") => "read".to_string(),
        _ => String::new(),
    }
}

fn default_error_message(code: &str) -> &str {
    match code {
        "file_not_found" => "file not found",
        "path_is_directory" => "path is directory",
        "file_too_large" => "file exceeds read limit",
        _ => "terminal resource operation failed",
    }
}

async fn collect_stdout(stream: Option<impl futures::Stream<Item = String> + Unpin>) -> String {
    match stream {
        Some(mut s) => {
            let mut buf = String::new();
            while let Some(chunk) = s.next().await {
                buf.push_str(&chunk);
            }
            buf
        }
        None => String::new(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_normalize_action_defaults_to_validate() {
        assert_eq!(normalize_action(None), "validate");
        assert_eq!(normalize_action(Some("")), "validate");
        assert_eq!(normalize_action(Some("  ")), "validate");
    }

    #[test]
    fn test_normalize_action_validate() {
        assert_eq!(normalize_action(Some("validate")), "validate");
        assert_eq!(normalize_action(Some("VALIDATE")), "validate");
        assert_eq!(normalize_action(Some(" Validate ")), "validate");
    }

    #[test]
    fn test_normalize_action_read() {
        assert_eq!(normalize_action(Some("read")), "read");
        assert_eq!(normalize_action(Some("READ")), "read");
    }

    #[test]
    fn test_normalize_action_invalid() {
        assert_eq!(normalize_action(Some("delete")), "");
        assert_eq!(normalize_action(Some("write")), "");
    }

    #[test]
    fn test_default_error_message_known_codes() {
        assert_eq!(default_error_message("file_not_found"), "file not found");
        assert_eq!(default_error_message("path_is_directory"), "path is directory");
        assert_eq!(default_error_message("file_too_large"), "file exceeds read limit");
    }

    #[test]
    fn test_default_error_message_unknown_code() {
        assert_eq!(
            default_error_message("something_else"),
            "terminal resource operation failed"
        );
    }
}

File: worker/worker-docker/internal/runner/terminal_resource.go
package runner

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
)

const (
	terminalResourceCapabilityName     = "terminalresource"
	terminalResourceCapabilityDeclared = "terminalResource"
	terminalResourceActionValidate     = "validate"
	terminalResourceActionRead         = "read"
	terminalResourceActionExport       = "export"
	terminalResourceCodeFileNotFound   = "file_not_found"
	terminalResourceCodePathIsDir      = "path_is_directory"
	terminalResourceCodeFileTooLarge   = "file_too_large"
)

type terminalResourcePayload struct {
	SessionID string `json:"session_id"`
	FilePath  string `json:"file_path"`
	Action    string `json:"action,omitempty"`
	SignedURL string `json:"signed_url,omitempty"`
}

type terminalResourceRequest struct {
	SessionID string
	FilePath  string
	Action    string
	SignedURL string
}

type terminalResourceRunResult struct {
	SessionID string `json:"session_id"`
	FilePath  string `json:"file_path"`
	MIMEType  string `json:"mime_type"`
	SizeBytes int64  `json:"size_bytes"`
	Blob      []byte `json:"blob,omitempty"`
}

type terminalResourceProbeResult struct {
	Error    string `json:"error,omitempty"`
	Message  string `json:"message,omitempty"`
	MIMEType string `json:"mime_type,omitempty"`
	Size     int64  `json:"size_bytes"`
	Blob     string `json:"blob,omitempty"`
}

var httpPutFile = putFileToSignedURL

const terminalResourceProbeScript = `
import argparse
import base64
import json
import mimetypes
import os
import sys

parser = argparse.ArgumentParser()
parser.add_argument("--action", choices=["validate", "read"], default="validate")
parser.add_argument("--file-path", required=True)
parser.add_argument("--max-read-bytes", type=int, required=True)
args = parser.parse_args()

target = args.file_path
if not os.path.exists(target):
    print(json.dumps({"error": "file_not_found", "message": "file not found"}))
    sys.exit(10)
if os.path.isdir(target):
    print(json.dumps({"error": "path_is_directory", "message": "path is directory"}))
    sys.exit(11)

size_bytes = os.path.getsize(target)
mime_type, _ = mimetypes.guess_type(target)
if not mime_type:
    mime_type = "application/octet-stream"

if args.action == "validate":
    print(json.dumps({"mime_type": mime_type, "size_bytes": size_bytes}))
    sys.exit(0)

limit = args.max_read_bytes
if size_bytes > limit:
    print(json.dumps({
        "error": "file_too_large",
        "message": "file exceeds read limit",
        "mime_type": mime_type,
        "size_bytes": size_bytes,
    }))
    sys.exit(12)

with open(target, "rb") as fh:
    content = fh.read(limit + 1)
if len(content) > limit:
    print(json.dumps({
        "error": "file_too_large",
        "message": "file exceeds read limit",
        "mime_type": mime_type,
        "size_bytes": len(content),
    }))
    sys.exit(12)

print(json.dumps({
    "mime_type": mime_type,
    "size_bytes": len(content),
    "blob": base64.b64encode(content).decode("ascii"),
}))
`

func (m *terminalSessionManager) ResolveResource(ctx context.Context, req terminalResourceRequest) (terminalResourceRunResult, error) {
	if m == nil {
		return terminalResourceRunResult{}, newTerminalExecError("execution_failed", terminalExecNotReadyMessage)
	}

	sessionID := strings.TrimSpace(req.SessionID)
	filePath := strings.TrimSpace(req.FilePath)
	if sessionID == "" || filePath == "" {
		return terminalResourceRunResult{}, newTerminalExecError(terminalExecCodeInvalidPayload, "session_id and file_path are required")
	}

	action := normalizeTerminalResourceAction(req.Action)
	if action == "" {
		return terminalResourceRunResult{}, newTerminalExecError(terminalExecCodeInvalidPayload, "action must be validate, read, or export")
	}
	signedURL := strings.TrimSpace(req.SignedURL)
	if action == terminalResourceActionExport && signedURL == "" {
		return terminalResourceRunResult{}, newTerminalExecError(terminalExecCodeInvalidPayload, "signed_url is required for export")
	}

	m.mu.Lock()
	session, ok := m.sessions[sessionID]
	if !ok || session == nil {
		m.mu.Unlock()
		return terminalResourceRunResult{}, newTerminalExecError(terminalExecCodeSessionNotFound, terminalExecNoSessionMessage)
	}
	if session.busy {
		m.mu.Unlock()
		return terminalResourceRunResult{}, newTerminalExecError(terminalExecCodeSessionBusy, terminalExecBusyMessage)
	}
	session.busy = true
	containerName := session.containerName
	m.mu.Unlock()

	resourceResult, err := m.resolveResourceInSession(ctx, sessionID, containerName, filePath, action, signedURL)
	if err != nil {
		if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
			m.destroySession(sessionID)
			return terminalResourceRunResult{}, err
		}
		var terminalErr *terminalExecError
		if errors.As(err, &terminalErr) && terminalErr.Code() == terminalExecCodeSessionNotFound {
			m.destroySession(sessionID)
			return terminalResourceRunResult{}, err
		}
		m.markSessionIdle(sessionID)
		return terminalResourceRunResult{}, err
	}

	if _, ok := m.markSessionIdle(sessionID); !ok {
		return terminalResourceRunResult{}, newTerminalExecError(terminalExecCodeSessionNotFound, terminalExecNoSessionMessage)
	}
	return resourceResult, nil
}

func (m *terminalSessionManager) resolveResourceInSession(
	ctx context.Context,
	sessionID string,
	containerName string,
	filePath string,
	action string,
	signedURL string,
) (terminalResourceRunResult, error) {
	probeAction := action
	if action == terminalResourceActionExport {
		probeAction = terminalResourceActionValidate
	}

	probe, err := probeTerminalResource(ctx, containerName, filePath, probeAction, m.outputLimitBytes)
	if err != nil {
		return terminalResourceRunResult{}, err
	}

	result := buildTerminalResourceResult(sessionID, filePath, probe)
	if action == terminalResourceActionExport {
		if err := exportTerminalResource(ctx, containerName, filePath, signedURL); err != nil {
			return terminalResourceRunResult{}, err
		}
		return result, nil
	}
	if action != terminalResourceActionRead {
		return result, nil
	}

	blobValue := strings.TrimSpace(probe.Blob)
	if blobValue == "" {
		result.Blob = []byte{}
		return result, nil
	}
	decoded, err := base64.StdEncoding.DecodeString(blobValue)
	if err != nil {
		return terminalResourceRunResult{}, fmt.Errorf("decode resource blob: %w", err)
	}
	result.Blob = decoded
	return result, nil
}

func probeTerminalResource(
	ctx context.Context,
	containerName string,
	filePath string,
	action string,
	maxReadBytes int,
) (terminalResourceProbeResult, error) {
	execResult := runDockerCommand(ctx, terminalExecDockerResourceArgs(containerName, action, filePath, maxReadBytes)...)
	if execResult.Err != nil {
		return terminalResourceProbeResult{}, fmt.Errorf("docker exec failed: %w", execResult.Err)
	}
	if isNoSuchContainerMessage(execResult.Stderr) {
		return terminalResourceProbeResult{}, newTerminalExecError(terminalExecCodeSessionNotFound, terminalExecNoSessionMessage)
	}

	if execResult.ExitCode != 0 {
		if probe, ok := tryDecodeTerminalResourceProbeOutput(execResult.Stdout); ok {
			if code := strings.TrimSpace(probe.Error); code != "" {
				message := terminalResourceErrorMessage(code, probe.Message)
				return terminalResourceProbeResult{}, newTerminalExecError(code, message)
			}
		}
		return terminalResourceProbeResult{}, fmt.Errorf(
			"docker exec failed: %s",
			terminalResourceDockerFailureMessage("exit code", execResult.ExitCode, execResult.Stderr, execResult.Stdout),
		)
	}

	probe, err := decodeTerminalResourceProbeOutput(execResult.Stdout)
	if err != nil {
		return terminalResourceProbeResult{}, fmt.Errorf("invalid terminalResource result: %w", err)
	}
	if code := strings.TrimSpace(probe.Error); code != "" {
		message := terminalResourceErrorMessage(code, probe.Message)
		return terminalResourceProbeResult{}, newTerminalExecError(code, message)
	}
	return probe, nil
}

func exportTerminalResource(ctx context.Context, containerName string, filePath string, signedURL string) error {
	tempFile, err := os.CreateTemp("", "onlyboxes-export-*")
	if err != nil {
		return fmt.Errorf("create temp file: %w", err)
	}
	tempPath := tempFile.Name()
	if closeErr := tempFile.Close(); closeErr != nil {
		os.Remove(tempPath)
		return fmt.Errorf("close temp file: %w", closeErr)
	}
	defer os.Remove(tempPath)

	copyResult := runDockerCommand(ctx, terminalExecDockerCopyArgs(containerName, filePath, tempPath)...)
	if copyResult.Err != nil {
		return fmt.Errorf("docker cp failed: %w", copyResult.Err)
	}
	if isNoSuchContainerMessage(copyResult.Stderr) {
		return newTerminalExecError(terminalExecCodeSessionNotFound, terminalExecNoSessionMessage)
	}
	if copyResult.ExitCode != 0 {
		return fmt.Errorf(
			"docker cp failed: %s",
			dockerCommandFailureMessage("exit code", copyResult.ExitCode, copyResult.Stderr),
		)
	}
	if err := httpPutFile(ctx, signedURL, tempPath); err != nil {
		return err
	}
	return nil
}

func putFileToSignedURL(ctx context.Context, signedURL string, filePath string) error {
	file, err := os.Open(filePath)
	if err != nil {
		return fmt.Errorf("open export file: %w", err)
	}
	defer file.Close()

	stat, err := file.Stat()
	if err != nil {
		return fmt.Errorf("stat export file: %w", err)
	}
	request, err := http.NewRequestWithContext(ctx, http.MethodPut, signedURL, file)
	if err != nil {
		return fmt.Errorf("build upload request: %w", err)
	}
	request.ContentLength = stat.Size()

	response, err := http.DefaultClient.Do(request)
	if err != nil {
		return fmt.Errorf("upload export file: %w", err)
	}
	defer response.Body.Close()
	if response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusMultipleChoices {
		body, _ := io.ReadAll(io.LimitReader(response.Body, 1024))
		message := strings.TrimSpace(string(body))
		if message == "" {
			message = response.Status
		}
		return fmt.Errorf("upload export file failed: %s", message)
	}
	return nil
}

func buildTerminalResourceResult(sessionID string, filePath string, probe terminalResourceProbeResult) terminalResourceRunResult {
	mimeType := strings.TrimSpace(probe.MIMEType)
	if mimeType == "" {
		mimeType = "application/octet-stream"
	}
	return terminalResourceRunResult{
		SessionID: sessionID,
		FilePath:  filePath,
		MIMEType:  mimeType,
		SizeBytes: probe.Size,
	}
}

func terminalExecDockerResourceArgs(containerName string, action string, filePath string, maxReadBytes int) []string {
	limit := maxReadBytes
	if limit <= 0 {
		limit = 1
	}
	return []string{
		"exec",
		containerName,
		"python3",
		"-c",
		terminalResourceProbeScript,
		"--action",
		action,
		"--file-path",
		filePath,
		"--max-read-bytes",
		strconv.Itoa(limit),
	}
}

func terminalExecDockerCopyArgs(containerName string, filePath string, targetPath string) []string {
	return []string{
		"cp",
		containerName + ":" + filePath,
		targetPath,
	}
}

func normalizeTerminalResourceAction(action string) string {
	switch strings.TrimSpace(strings.ToLower(action)) {
	case "":
		return terminalResourceActionValidate
	case terminalResourceActionValidate:
		return terminalResourceActionValidate
	case terminalResourceActionRead:
		return terminalResourceActionRead
	case terminalResourceActionExport:
		return terminalResourceActionExport
	default:
		return ""
	}
}

func decodeTerminalResourceProbeOutput(stdout string) (terminalResourceProbeResult, error) {
	trimmed := strings.TrimSpace(stdout)
	if trimmed == "" {
		return terminalResourceProbeResult{}, errors.New("empty output")
	}
	decoded := terminalResourceProbeResult{}
	if err := json.Unmarshal([]byte(trimmed), &decoded); err != nil {
		return terminalResourceProbeResult{}, err
	}
	return decoded, nil
}

func tryDecodeTerminalResourceProbeOutput(stdout string) (terminalResourceProbeResult, bool) {
	decoded, err := decodeTerminalResourceProbeOutput(stdout)
	if err != nil {
		return terminalResourceProbeResult{}, false
	}
	return decoded, true
}

func terminalResourceDockerFailureMessage(prefix string, value int, stderr string, stdout string) string {
	message := dockerCommandFailureMessage(prefix, value, stderr)
	if strings.TrimSpace(stderr) != "" {
		return message
	}

	trimmedStdout := strings.TrimSpace(stdout)
	if trimmedStdout == "" {
		return message
	}
	if len(trimmedStdout) > 256 {
		trimmedStdout = trimmedStdout[:256] + "..."
	}
	return message + ", stdout=" + trimmedStdout
}

func terminalResourceErrorMessage(code string, fallback string) string {
	if trimmed := strings.TrimSpace(fallback); trimmed != "" {
		return trimmed
	}
	switch strings.TrimSpace(code) {
	case terminalResourceCodeFileNotFound:
		return "file not found"
	case terminalResourceCodePathIsDir:
		return "path is directory"
	case terminalResourceCodeFileTooLarge:
		return "file exceeds read limit"
	default:
		return "terminal resource operation failed"
	}
}
